### PR TITLE
[Experiment] Replace crate-hash with simple timestamp.

### PR DESCRIPTION
**Pull request by michaelwoerister**
This might be a nice simplification that make things more robust and would probably also simplify any work on end-to-end queries. Let's see what performance looks like.

r? @ghost

### CODE REVIEW

**Comment by michaelwoerister** [comment]
It looks like it's not that trivial to make this pass all tests. But let's see if we can get a build for perf.rlo.

@bors try @rust-timer queue

**Comment by bjorn3** [comment]
This would make builds non-reproducible and if anything increase the risk of multiple crates having the same crate hash due to being either built very close to each other on the same machine, multiple machines having the same time or the time going backwards due to eg an empty cmos battery or ntp adjustments.

**Comment by michaelwoerister** [comment]
Sure, this is just an experiment to see the maximum impact something like this can have on incr comp performance. Reproducible builds are high priority. An actual implementation would have to take that into account.

I'm starting to wonder though if we need the crate hash in the current form at all.

**Comment by bjorn3** [comment]
> I'm starting to wonder though if we need the crate hash in the current form at all.

The crate hash is what ensures that replacing a dependency always forces the dependent crate to be recompiled. (Something more finegrained may make it impossible to eg add extra query dependencies without causing breakage) It is also essential for incr comp to invalidate queries based on changed crate metadata.

**Comment by cjgillot** [comment]
I am not sure this perf run will measure what we want it to measure. 
In my understanding, one of the main roles of crate hash is to tell the query system whether we should load a value from metadata. This is useful to skip loading values from _upstream_ crates that are known to be green. Changing to a timestamp will not change the behaviour: the perf infra does not rebuild upstream crates, so their crate hash won't change between to builds, and we will only measure the cost of computing the hash. 
If we want to measure the effect of always loading data, we had better mark crate hash query as no_hash. This will prevent the query system from using its value to mark stuff green, and force loading the data from disk.

File: compiler/rustc_hir/src/hir.rs [hunk 1]
```diff
@@ -776,7 +776,6 @@ impl<T> MaybeOwner<T> {
 #[derive(Debug)]
 pub struct Crate<'hir> {
     pub owners: IndexVec<LocalDefId, MaybeOwner<&'hir OwnerInfo<'hir>>>,
-    pub hir_hash: Fingerprint,
```
Comment (cjgillot) [inline compiler/rustc_hir/src/hir.rs:779]: Why this change? This may force re-executions that are unrelated to `crate_hash`.

**Comment by michaelwoerister** [comment]
Let me try to elaborate where this might be going. Some background first: With respect to incremental compilation, the crate hash is used to determine if something from crate metadata can have changed at all. I.e. it is the root dependency node for all queries that read from the metadata of a given upstream crate. The requirement for it is thus that if anything in crate metadata changes, it too needs to reflect that change.

Right now crate hash is computed from a set of things that supposedly makes it fulfill this requirement: the HIR, (some) commandline args, the file names of the current crate's source files after remapping, etc. There are some problems with this approach though:

 - We manually put together the list of everything that goes into this hash. This is error-prone as it is hard to tell what should go into it and what shouldn't. @cjgillot recently found a case where we miss changes to upstream source files, for example. Commandline arguments are another example, where each argument has to be tagged and it is easy to miss something.
 - Among other things we construct the crate hash from the HIR, which already has made querifying HIR-related things more complicated. Not catastrophically so, but it is one more thing that has to be taken into account when working in this area. The situation has improved quite a bit since @cjgillot's recent refactorings there, I think, but in the past the attempt to avoid hashing things twice has been a major source of bugs and hard-to-understand code.
 - IIUC, right now, in non-incremental mode, we are hashing the entire HIR just for the crate hash. I suspect that this has a small but measurable impact on compile times.

So, what I'm trying to figure out is: Can we make this cheaper, simpler, and more robust at the same time.

I see several options:

 1. Remove the crate hash entirely and just make the root dep-node for crate metadata `no_hash` as @cjgillot says above. 
  - Pro: we can remove a bunch of complicated and costly computations entirely. Nothing can go wrong from a correctness standpoint.
  - Con: we force the compiler to re-fingerprint everything coming from crate metadata, even if there is no change at all, which is a rather common case. 

 2. Compute the crate hash by simply hashing the bytes of crate metadata.
  - Pro: It is much simpler than the current approach. No other part of compiler has to know about it.
  - Pro: We cannot miss any data from the current crate, even if arbitrary new stuff gets added to the compiler. We don't need to try and come up with an exhaustive list of things that we think can have an influence.
  - Pro: It's probably cheaper computationally because bulk-hashing flat byte streams is something that computers are really good at. This would be even more true if we switched to something like UMASH fingerprints or XXH3 for hashing -- but that wouldn't be a necessary prerequisite, I suspect.
  - Con: It might be slightly more sensitive to changes that don't make a semantic difference. However, the current implementation already is very sensitive and the most important use case, where Cargo doesn't even touch the upstream crate would be just the same as it is now.
  - Con: I'm not a 100% certain if we also need to include the crate hashes of upstream crates in order to catch all possible cases (like we do now). But worst case we just do that.

 3. We compute the crate hash by hashing source file contents instead of HIR. This would allow us to decouple the crate hash from the HIR, which can only be good.
  - Pro: Much simpler than hashing the HIR, especially on the implementation side. Makes HIR related refactorings easier.
  - Con: Sensitive to semantically irrelevant changes.
  - Con: We have to account for the results of macro expansion too somehow. This might be as simple as including the crate hash of proc-macro crates -- but that would assume that proc-macros are deterministic which in general they are not required to be. 
  - Con: We still have to feed all the other things like command-line arguments into the hash. So this approach inherits most of the robustness problems of the current approach.

I think that option (2) is by far the most promising. Option (1) is a distant second because of the performance implications. Option (3) might be even worse than the current approach due to the macro problem.

Consider yourselves invited to try and poke holes in my reasoning `:)`

File: compiler/rustc_hir/src/hir.rs (hunk 1 above)
Comment (michaelwoerister) [reply compiler/rustc_hir/src/hir.rs:779]: Yes, that's true, making the `hir_crate` query `no_hash` is an unrelated change. Although from the performance results it looks like it does not make much of difference?

**Comment by cjgillot** [comment]
I agree with your reasoning.

> I see several options:
> 
>     1. Remove the crate hash entirely and just make the root dep-node for crate metadata `no_hash` as @cjgillot says above.
> 
> 
>     * Pro: we can remove a bunch of complicated and costly computations entirely. Nothing can go wrong from a correctness standpoint.
> 
>     * Con: we force the compiler to re-fingerprint everything coming from crate metadata, even if there is no change at all, which is a rather common case.
> 

Remark: we already partially re-fingerprint data we are loading from disk even when the `crate_hash` is green, as part of the "unstable fingerprint" check.

> 
>     2. Compute the crate hash by simply hashing the bytes of crate metadata.
> 
> 
>     * Pro: It is much simpler than the current approach. No other part of compiler has to know about it.
> 
>     * Pro: We cannot miss any data from the current crate, even if arbitrary new stuff gets added to the compiler. We don't need to try and come up with an exhaustive list of things that we think can have an influence.
> 
>     * Pro: It's probably cheaper computationally because bulk-hashing flat byte streams is something that computers are really good at. This would be even more true if we switched to something like UMASH fingerprints or XXH3 for hashing -- but that wouldn't be a necessary prerequisite, I suspect.
> 
>     * Con: It might be slightly more sensitive to changes that don't make a semantic difference. However, the current implementation already is very sensitive and the most important use case, where Cargo doesn't even touch the upstream crate would be just the same as it is now.
> 
>     * Con: I'm not a 100% certain if we also need to include the crate hashes of upstream crates in order to catch all possible cases (like we do now). But worst case we just do that.
> 

I just posted https://github.com/rust-lang/rust/issues/94878 as a call for participation to investigate this. I am available to mentor it, you are welcome to join.

"Con 1" may just be the necessary cost for our peace of mind here. We will have to measure the perf effect of course.
"Con 2" do we need to include the upstream crate hash? If we hash exactly what goes into metadata, we may not need to.

One tricky thing will be to have the stable hasher and the metadata encoding agree:
- metadata encoder will see a `DefId` as a pair of integers, while the stable hasher will see it as a `DefPathHash`;
- metadata encoder is order-sensitive, while the stable hasher ignores the order of `DefId`s.

> 
>     3. We compute the crate hash by hashing source file contents instead of HIR. This would allow us to decouple the crate hash from the HIR, which can only be good.
> 
> 
>     * Pro: Much simpler than hashing the HIR, especially on the implementation side. Makes HIR related refactorings easier.
> 
>     * Con: Sensitive to semantically irrelevant changes.
> 
>     * Con: We have to account for the results of macro expansion too somehow. This might be as simple as including the crate hash of proc-macro crates -- but that would assume that proc-macros are deterministic which in general they are not required to be.
> 
>     * Con: We still have to feed all the other things like command-line arguments into the hash. So this approach inherits most of the robustness problems of the current approach.
> 

In general, proc macros are not deterministic, they could even emit a time stamp. Only regular macros are. In consequence, we need to hash the post-expansion AST, or to add the output of each proc-macro expansion to the crate hash.

> 
> I think that option (2) is by far the most promising. Option (1) is a distant second because of the performance implications. Option (3) might be even worse than the current approach due to the macro problem.
> 
> Consider yourselves invited to try and poke holes in my reasoning `:)`

I agree with the ordering of solutions, and I haven't found holes yet :).

**Comment by bjorn3** [comment]
> Con: I'm not a 100% certain if we also need to include the crate hashes of upstream crates in order to catch all possible cases (like we do now). But worst case we just do that.

Yes we do. In any case the crate metadata already contains the crate hash for all dependencies so hashing the crate metadata is enough.

Note that hashing the crate metadata is not enough for proc macros. The crate metadata can be unchanged despite the actual implementation changing. If the implementation changes the crate hash needs to change to ensure that we don't forget to expand the proc macro invocations again.

**Comment by michaelwoerister** [comment]
> metadata encoder will see a DefId as a pair of integers, while the stable hasher will see it as a DefPathHash;

That's the main uncertainty I see with this approach, yes. It might be OK to hash the bytes verbatim because the tables needed for decoding the indices are part of the same value - but I'm not certain that that covers everything. The approach in https://github.com/rust-lang/rust/issues/94878 should not have that issue though.

**Comment by cjgillot** [comment]
As a first step, we could reuse what @Aaron1011 did in https://github.com/rust-lang/rust/pull/87896 (= just sha256 the binary blob), and refine from there if the perf effect is too large.

**Comment by michaelwoerister** [comment]
Hashing the binary blob is what I suggested above. However, it does potentially have the problem you mentioned.

But if we also feed the crate hashes of all upstream crates into the hash, we will have transitively hashed all the data that can be used to do the defid->defpathhash mapping and thus should be able to detect any relevant change.

I haven't had time to really think through this though - so take it with a grain of salt :)

**Comment by michaelwoerister** [comment]
>> Con: I'm not a 100% certain if we also need to include the crate hashes of upstream crates in order to catch all possible cases (like we do now). But worst case we just do that.
>
> Yes we do. 

@bjorn3, can you elaborate your reasoning here? I don't disagree but I'd be interested in why you think it's necessary.

> In any case the crate metadata already contains the crate hash for all dependencies so hashing the crate metadata is enough.

Ah yes, that's true.

> Note that hashing the crate metadata is not enough for proc macros. The crate metadata can be unchanged despite the actual implementation changing. If the implementation changes the crate hash needs to change to ensure that we don't forget to expand the proc macro invocations again.

I assume you are referring to "This might be as simple as including the crate hash of proc-macro crates" in Option (3), right?

**Comment by bjorn3** [comment]
> @bjorn3, can you elaborate your reasoning here? I don't disagree but I'd be interested in why you think it's necessary.

Upstream crates can affect things like the layout of used types. Layouts are not stored in the crate metadata and thus wouldn't change the crate hash unless dependencies are included in the crate hash.

**Comment by bjorn3** [comment]
> I assume you are referring to "This might be as simple as including the crate hash of proc-macro crates" in Option (3), right?

I was refering to option 2. Hashing just the crate metadata would not contain enough data to uniquely identify proc macros when their implementation changes, which may cause proc macros to not be re-expanded on implementation changes if macro expansion becomes incremental.

**Comment by michaelwoerister** [comment]
I've been thinking a bit more about this. IMO, the only reason why hashing the crate metadata as a flat byte stream is not trivially correct is that the CrateLoader does some additional processing (like computing the cnum_map) which might introduce data dependencies on things that come from outside the byte stream. The way things are set up right now, it is very hard to tell what data goes into computing this additional data structures. 

If things were set up in a way that there was an eval_always query that only allowed access to the MetadataBlob then things would be simpler.

I think what @cjgillot proposes in https://github.com/rust-lang/rust/issues/94878 would be an improvement over what we have currently. I'd be very interested in seeing how it performs.

I'm going to close this PR as it's clear that we won't merge it.

**Comment by michaelwoerister** [comment]
> > @bjorn3, can you elaborate your reasoning here? I don't disagree but I'd be interested in why you think it's necessary.
> 
> Upstream crates can affect things like the layout of used types. Layouts are not stored in the crate metadata and thus wouldn't change the crate hash unless dependencies are included in the crate hash.

Wouldn't you necessarily also introduce a dependency on the crate hash of the upstream crate in such cases?

**Comment by michaelwoerister** [comment]
> > I assume you are referring to "This might be as simple as including the crate hash of proc-macro crates" in Option (3), right?
> 
> I was refering to option 2. Hashing just the crate metadata would not contain enough data to uniquely identify proc macros when their implementation changes, which may cause proc macros to not be re-expanded on implementation changes if macro expansion becomes incremental.

So this would be in the case where the upstream crate is a proc-macro?

**Comment by bjorn3** [comment]
At the query graph level yes, all data loaded from dependencies results in a dependency on the crate_hash query. At the crate metadata level, the explicit list of the dependency in the crate metadata is the only thing ensuring that dependency versions aren't mismatched. This list needs to be part of the crate hash one way or another. Hashing just file sources would not be enough for this. Hashing full crate metadata would be enough.

**Comment by michaelwoerister** [comment]
Right, proc-macro crates don't seem to encode their MIR at the moment -- so I can see how this is a problem.

**Comment by michaelwoerister** [comment]
Hm, given that proc-macros are not required to be deterministic, they would need to be handled specially anyway, I guess. I.e. even a hash that completely captured the code of the proc macro would not be strictly correct since the proc-macro can still pull in data from arbitrary sources. So we would have to unconditionally assume that it has changed.

**Comment by bjorn3** [comment]
For something like rust-analyzer unconditionally assuming that proc macro output has changed would result in 100% cpu usage if any proc macro is used. It should re-expand exactly when the proc macro needs to be/has been recompiled. In the future when rust-analyzer and rustc share more code it would be nice to be able to use the crate hash for that purpose.
